content/blog/2019-10-25-efficient-path-profiling/index.md

```diff
@@ -0,0 +1,109 @@
+Specifically, we compute the maximum spanning tree with respect to edge weights.
+All updates then only reason about those edges in the _chord_, the set of edges not in the maximum spanning tree.
+
+Instrumentation for the graph (choosing which register each path updates) is then added for each edge in the chord based on the integers computed earlier.
+Since each path of integers must be unique, the register selected by a given chord edge is simply given as the minimum integer increment since the last chord edge for any path.
+This algorithm gives the instrumentation for the integers calculated on our sample CFG as follows, where edges with squares are part of the chord:
+
+<img src="instrumentation.png" alt="drawing" width="300"/>
+
+### Regenerating Paths
+
+Finally, after incrementing each register according to our instrumentation, we must recover the number of times each path was taken.
+The integers calculated earlier give an intuition for reconstructing these values, since each integer path must have been unique.
+We can then just calculate the integer value associated with each path by walking the path in the integer version of the CFG and use this to reference which register corresponds to each path.
+
+### Extending to Cycles
+
+We have assumed so far that every graph is acyclic; how can we extend this algorithm to work with cycles?
+It turns out that evaluating each type of cycle and how it relates to an arbitrarily inserted enter or exit node provides sufficient information to construct our integer edge values as with a DAG.
+
+Cycles added to a DAG can be thought of as _backedges_, edges which visit a node previously seen in the DAG.
+Through replacing these backedges with information-carrying forward edges, the algorithm reconstructs a DAG from any cyclic graph.
+The details of these edges rely on technical casework, so those interested in the details should read the EPP paper directly; however, it is sufficient for our overview to state that each backedge is replaced by forward edges from the entry node to its target, and from its source to the exit node.
+This process can be summarized visually with this excellent diagram:
+
+<img src="cycle.png" alt="drawing" width="300"/>
+
+## Evaluation
+
+To evaluate how realistic this approach can be, the authors built the PP tool and compared it with an existing path profiling tool (QPT2).
+Several optimizations were applied to PP to promote a realistic comparison, such as mapping local registers through the Executable Editing Library (EEL) and implementing hash tables to handle a large number of paths.
+All experiments were run on the SPEC95 benchmark, which consists of a standard set of C and Fortran programs.
+
+Experiments were run on a Sun Ultraserver, with 167Mhz UltraSPARC processors and a whopping 2GB of memory.
+PP's overhead on this test suite averaged 30.9%, while QPT2's overhead averaged 16.1%.
```

Comment:
Hmm… but above, I thought you said QPT2 was the existing path profiling baseline?

> an existing path profiling tool (QPT2).

Perhaps you meant an existing edge profiling tool? Or perhaps just clarify that QPT2 is a "fake" path profiler that actually just does edge profiling and incorrectly assumes that it can infer the paths from that, as discussed above.

Reply:
Ah, the earlier comment was wrong; QPT2 is an edge profiling tool.  Corrected, thanks!